src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp

```diff
@@ -351,58 +337,51 @@ namespace OpenMS
       dist = f.getIsotopeDistribution(FineIsotopePatternGenerator(max_isotope_probability_));
     }
 
-    String ion_name = String(Residue::residueTypeToIonLetter(res_type)) + String(ion.size()) + String((Size)abs(charge), '+');
-
     for (const auto& it : dist)
     {
-      p.setMZ(it.getMZ() / charge);
-      p.setIntensity(intensity * it.getIntensity());
       if (add_metainfo_) // one entry per peak
       {
         ion_names.push_back(ion_name);
         charges.push_back(charge);
       }
-      spectrum.push_back(p);
+      spectrum.emplace_back(it.getMZ() / charge, intensity * it.getIntensity());
     }
   }
 
   void addLosses_faster_(PeakSpectrum& spectrum,
                          double mz,
-                         std::set<EmpiricalFormula>& f_losses,
+                         const std::set<EmpiricalFormula>& f_losses,
                          int ion_ordinal,
                          DataArrays::StringDataArray& ion_names,
                          DataArrays::IntegerDataArray& charges,
                          double intensity,
-                         Residue::ResidueType res_type,
+                         const Residue::ResidueType res_type,
                          bool add_metainfo,
                          int charge)
   {
-    std::vector<double> losses;
-    std::vector<String> losses_names;
-    {
-      for (const auto& formula : f_losses)
-      {
-        losses.push_back( formula.getMonoWeight() );
-        if (add_metainfo) 
-        {
-          losses_names.emplace_back(formula.toString());
-        }
-      }
-    }
+    const String charge_str((Size)abs(charge), '+');
+    const String residue_str(Residue::residueTypeToIonLetter(res_type));
+    const String ion_ordinal_str(String(ion_ordinal) + "-");
+
+    // formula.toString() is extremely expensive, so we use a static map to remember what String belongs to which formula
+    static std::map<EmpiricalFormula, String> formula_str_cache;
```

Comment:
the cache should be precomputed and passed to the function. please no static function variables if it can be avoided

Reply:
I've made the map a member of the TSG, so it is not static anymore. Precomputing seems rather complicated, as the set fx_losses is built up continously before every execution of addLossesFaster_().

Reply:
seems still static here.

Reply:
Just pushed the commit :)

Reply:
I think this might not work in a multi-threaded context

Reply:
True. I am gonna have a look at how to precompute the map, so it can be thread-safe.

Reply:
sorry for giving so much headache. Most of my complaints are because of the non-ideal original design

Reply:
and not because of your (by the way highly appreciated) changes

Reply:
Still different results with different threads, but I tested a develop-build and actually there, the results vary too. So maybe, we should take a look into the algorithm in general.

Reply:
Confirmed. The SimpleSearchEngine has different outputs for different threadnums in the current develop. The differences are very slight. Also, e.g. with four threads, the results vary between two runs.